dhctl/pkg/config/validation_rules.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2021 Flant JSC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"sigs.k8s.io/yaml"
+)
+
+type ValidationRule func(oldValue, newValue json.RawMessage) error
+
+type RuleValidator struct {
+	rules      map[string]ValidationRule
+	validators map[string]*RuleValidator
+}
+
+func (v *RuleValidator) CreateRule(path string, rule ValidationRule) {
+	separatorIndex := strings.IndexByte(path, '.')
+	if separatorIndex == -1 {
+		if v.rules == nil {
+			v.rules = make(map[string]ValidationRule)
+		}
+		v.rules[path] = rule
+		return
+	}
+
+	if v.validators == nil {
+		v.validators = make(map[string]*RuleValidator)
+	}
+
+	if v.validators[path[:separatorIndex]] == nil {
+		v.validators[path[:separatorIndex]] = &RuleValidator{}
+	}
+
+	v.validators[path[:separatorIndex]].CreateRule(path[separatorIndex+1:], rule)
+}
+
+func NewDefaultRuleValidators() map[SchemaIndex]RuleValidator {
+	var v RuleValidator
+	v.CreateRule("masterNodeGroup.replicas", NotLessRule)
+	v.CreateRule("nodeGroups.replicas", NotLessRule)
```

Comment:
1. replicas in cloud permanent node groups can be 0. For example, i want remove nodes and save node group
2. I don’t quite understand how this will work node groups is an array

Reply:
nodeGroups.replicas validation here is our error, there should be no validation of nodeGroups.replicas. This will be fixed soon.